[Pointer Events] Implement the `auxclick` event

rr-codes opened the pull request:
#### 62f87d3a9254d50d77d0feefab835c422c55e7eb
<pre>
[Pointer Events] Implement the `auxclick` event
<a href="https://bugs.webkit.org/show_bug.cgi?id=22382">https://bugs.webkit.org/show_bug.cgi?id=22382</a>
<a href="https://rdar.apple.com/25988904">rdar://25988904</a>

Reviewed by Abrar Rahman Protyasha and Darin Adler.

Implement the `auxclick` UI event as per the corresponding UI events specification (<a href="https://w3c.github.io/uievents/#event-type-auxclick).">https://w3c.github.io/uievents/#event-type-auxclick).</a>

Also adds a runtime feature flag to control its enablement.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::eventListenersDidChange):
* Source/WebCore/dom/Element.cpp:
(WebCore::isCompatibilityMouseEvent):
(WebCore::dispatchPointerEventIfNeeded):
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::canTriggerActivationBehavior):
* Source/WebCore/dom/mac/ImageControlsMac.cpp:
(WebCore::ImageControlsMac::handleEvent):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::updateSaveButton):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::willDispatchEvent):
(WebCore::HTMLInputElement::defaultEventHandler):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::defaultEventHandler):
* Source/WebCore/html/ImageDocument.cpp:
(WebCore::ImageEventListener::handleEvent):
* Source/WebCore/html/MediaDocument.cpp:
(WebCore::MediaDocument::defaultEventHandler):
* Source/WebCore/html/shadow/AutoFillButtonElement.cpp:
(WebCore::AutoFillButtonElement::defaultEventHandler):
* Source/WebCore/html/shadow/DataListButtonElement.cpp:
(WebCore::DataListButtonElement::defaultEventHandler):
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::defaultEventHandler):
* Source/WebCore/mathml/MathMLSelectElement.cpp:
(WebCore::MathMLSelectElement::defaultEventHandler):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseReleaseEvent):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::triggerOptionalStorageAccessQuirk const):

Canonical link: <a href="https://commits.webkit.org/281169@main">https://commits.webkit.org/281169@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12c6d306c6daff0a7e92148488525ba047fbf31e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11492 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62643 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9443 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61129 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45979 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9646 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47717 "") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6738 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51042 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28575 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32582 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8324 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8447 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52092 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54533 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8604 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64340 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58241 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8560 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55038 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2922 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51061 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55141 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2447 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80002 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8816 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34157 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13840 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35241 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36326 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34987 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
All (most?) of the test expectation mismatches surfaced by EWS are progressions -- please to update!

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -450,7 +450,7 @@ static ShouldIgnoreMouseEvent dispatchPointerEventIfNeeded(Element& element, con
         auto& pointerCaptureController = page->pointerCaptureController();
 #if ENABLE(TOUCH_EVENTS)
-        if (platformEvent.pointerId() != mousePointerID && mouseEvent.type() != eventNames().clickEvent && pointerCaptureController.preventsCompatibilityMouseEventsForIdentifier(platformEvent.pointerId()))
+        if (platformEvent.pointerId() != mousePointerID && mouseEvent.type() != eventNames().clickEvent && mouseEvent.type() != eventNames().auxclickEvent && pointerCaptureController.preventsCompatibilityMouseEventsForIdentifier(platformEvent.pointerId()))
```
Should we make a helper function that checks for both `click` and `auxclick` so we don’t have to make the expression longer in each place? I see 15 places we could use it.

rr-codes replied:
definitely; would a helper method called `isClickOrAuxClickEvent` in `Event.h` be good? Or is there a better place for it?

rr-codes replied:
(I ended up going with `static bool isClickOrAuxClickEvent(const Event&);` on MouseEvent.h

darinadler commented on Source/WebCore/Modules/notifications/Notification.cpp:
```diff
@@ -426,6 +426,7 @@ void Notification::requestPermission(Document& document, RefPtr<NotificationPerm
 void Notification::eventListenersDidChange()
 {
     m_hasRelevantEventListener = hasEventListeners(eventNames().clickEvent)
+        || hasEventListeners(eventNames().auxclickEvent)
```
This was in alphabetical order before, so maybe we could put `auxclickEvent` on the top line before the `||`?

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2333,10 +2333,30 @@ HandleUserInputEventResult EventHandler::handleMouseReleaseEvent(const PlatformM
+        }
+
+        // The click event should only be fired for the primary pointer button.
+        return !dispatchMouseEvent(eventNames().clickEvent, nodeToClick.get(), m_clickCount, platformMouseEvent, FireMouseOverOut::Yes);
```
Seems like we should just put the event name in a local variable and share this line of code.

rr-codes replied:
I had it like this originally, but once I added the new if condition branch for the feature flag, that branch no longer dispatches an event like this, so the logic would become slightly more complicated if the event name were to be put in a local variable.

darinadler replied:
I understand, and yet I still think I might refactor it to share that line of code, given that we never dispatch 2 events, only 0 or 1. But I’m OK landing it like this.

rr-codes replied:
I'll make that change first, and also fix the tests too

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -510,7 +510,8 @@ Element::DispatchMouseEventResult Element::dispatchMouseEvent(const PlatformMous
 
-    if (mouseEvent->type() == eventNames().clickEvent && mouseEvent->detail() == 2) {
+    // The document should not receive dblclick for non-primary buttons.
+    if ((mouseEvent->type() == eventNames().clickEvent) && mouseEvent->detail() == 2) {
```
Nit: Undo the added parentheses?

aprotyas commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1287,7 +1287,7 @@ Quirks::StorageAccessResult Quirks::triggerOptionalStorageAccessQuirk(Element& e
     static NeverDestroyed<UserScript> kinjaLoginUserScript { "function triggerLoginForm() { let elements = document.getElementsByClassName('js_header-userbutton'); if (elements && elements[0]) { elements[0].click(); clearInterval(interval); } } let interval = setInterval(triggerLoginForm, 200);"_s, URL(aboutBlankURL()), Vector<String>(), Vector<String>(), UserScriptInjectionTime::DocumentEnd, Use ...
 
-    if (eventType == eventNames().clickEvent) {
+    if (eventType == eventNames().clickEvent || eventType == eventNames().auxclickEvent) {
```
Also use `MouseEvent::isClickOrAuxClickEvent()` here?

rr-codes replied:
I can't, since we only have access to the type and not the event

aprotyas replied:
Oh, I see. Maybe? Or not, up to you.

```cpp
bool MouseEvent::isClickOrAuxClickEvent(Event& e) { return MouseEvent::isClickOrAuxClickEventType(e.type()); }

bool MouseEvent::isClickOrAuxClickEventType(const AtomString& type)
{
    return type == eventNames().clickEvent || type == eventNames().auxclickEvent;
}
```

rr-codes replied:
I thought about this, but then realized it would require every other call site to use `event.type()` instead, which seemed slightly more inconvenient than keeping this one line

darinadler replied:
I think the overloading is a clever way to do both.

rr-codes replied:
ah, I didn't realize the suggestion was to use both methods!

darinadler commented on Source/WebCore/dom/MouseEvent.h:
```diff
@@ -85,6 +85,8 @@ class MouseEvent : public MouseRelatedEvent {
 
     static bool canTriggerActivationBehavior(const Event&);
+    static bool typeIsClickOrAuxClick(const Event&);
+    static bool typeIsClickOrAuxClick(const AtomString&);
```
Maybe a slightly more terse name. I like `isAnyClick`, `isEitherClick`, or `isClickOrAuxClick`.

Maybe these should be non-member functions? I don’t see why we have to make them members of `MouseEvent`. It’s good to have them in this header file, though!

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2264,6 +2264,31 @@ static RefPtr<Node> targetNodeForClickEvent(Node* mousePressNode, Node* mouseRel
+    auto nodeToClick = targetNodeForClickEvent(RefPtr { m_clickNode }.get(), mouseEvent.protectedTargetNode().get());
+
+    if (!nodeToClick)
+        return false;
```
I’d omit the blank line here.

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2264,6 +2264,31 @@ static RefPtr<Node> targetNodeForClickEvent(Node* mousePressNode, Node* mouseRel
+    bool isPrimaryPointerButton = platformMouseEvent.button() == MouseButton::Left;
+
+    if (!isPrimaryPointerButton && !protectedFrame()->settings().auxclickEventEnabled())
+        return false;
```
I’d omit the blank line here.

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2264,6 +2264,31 @@ static RefPtr<Node> targetNodeForClickEvent(Node* mousePressNode, Node* mouseRel
+
+    const auto& eventName = isPrimaryPointerButton ? eventNames().clickEvent : eventNames().auxclickEvent;
+
+    return !dispatchMouseEvent(eventName, nodeToClick.get(), m_clickCount, platformMouseEvent, FireMouseOverOut::Yes);
```
I’d omit the blank line here. Also, just `auto&` is fine, don’t need to write `const auto&`.

darinadler commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -462,6 +462,8 @@ class EventHandler final : public CanMakeCheckedPtr<EventHandler> {
     enum class Cancelable : bool { No, Yes };
     bool dispatchMouseEvent(const AtomString& eventType, Node* target, int clickCount, const PlatformMouseEvent&, FireMouseOverOut);
 
+    bool swallowClickOrAuxclickEvent(const PlatformMouseEvent&, const MouseEventWithHitTestResults&);
```
Maybe `swallowAnyClickEvent` or `swallowEitherClickEvent`? Names without "and" seem to read more smoothly, even if they are a tad less precise.

darinadler commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1284,7 +1284,7 @@ Quirks::StorageAccessResult Quirks::triggerOptionalStorageAccessQuirk(Element& e
     static NeverDestroyed<UserScript> kinjaLoginUserScript { "function triggerLoginForm() { let elements = document.getElementsByClassName('js_header-userbutton'); if (elements && elements[0]) { elements[0].click(); clearInterval(interval); } } let interval = setInterval(triggerLoginForm, 200);"_s, URL(aboutBlankURL()), Vector<String>(), Vector<String>(), UserScriptInjectionTime::DocumentEnd, Use ...
 
-    if (eventType == eventNames().clickEvent) {
+    if (eventType == eventNames().clickEvent || eventType == eventNames().auxclickEvent) {
```
Still not using the helper function here yet!

darinadler reviewed (approved):
Some more ideas for tweaking style a bit.

rr-codes committed:
[Pointer Events] Implement the `auxclick` event
https://bugs.webkit.org/show_bug.cgi?id=22382
rdar://25988904

Reviewed by Abrar Rahman Protyasha and Darin Adler.

Implement the `auxclick` UI event as per the corresponding UI events specification (https://w3c.github.io/uievents/#event-type-auxclick).

Also adds a runtime feature flag to control its enablement.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::eventListenersDidChange):
* Source/WebCore/dom/Element.cpp:
(WebCore::isCompatibilityMouseEvent):
(WebCore::dispatchPointerEventIfNeeded):
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::canTriggerActivationBehavior):
* Source/WebCore/dom/mac/ImageControlsMac.cpp:
(WebCore::ImageControlsMac::handleEvent):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::updateSaveButton):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::willDispatchEvent):
(WebCore::HTMLInputElement::defaultEventHandler):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::defaultEventHandler):
* Source/WebCore/html/ImageDocument.cpp:
(WebCore::ImageEventListener::handleEvent):
* Source/WebCore/html/MediaDocument.cpp:
(WebCore::MediaDocument::defaultEventHandler):
* Source/WebCore/html/shadow/AutoFillButtonElement.cpp:
(WebCore::AutoFillButtonElement::defaultEventHandler):
* Source/WebCore/html/shadow/DataListButtonElement.cpp:
(WebCore::DataListButtonElement::defaultEventHandler):
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
(WebCore::SearchFieldCancelButtonElement::defaultEventHandler):
* Source/WebCore/mathml/MathMLSelectElement.cpp:
(WebCore::MathMLSelectElement::defaultEventHandler):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseReleaseEvent):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::triggerOptionalStorageAccessQuirk const):

Canonical link: https://commits.webkit.org/281169@main
